lib/mysql_connection.cpp

```diff
@@ -563,13 +571,25 @@ void MySQL_Connection::connect_start() {
 		mysql_ssl_set(mysql, mysql_thread___ssl_p2s_key, mysql_thread___ssl_p2s_cert, mysql_thread___ssl_p2s_ca, NULL, mysql_thread___ssl_p2s_cipher);
 	}
 	unsigned int timeout= 1;
+	const char *csname = NULL;
 	mysql_options(mysql, MYSQL_OPT_CONNECT_TIMEOUT, (void *)&timeout);
-	const MARIADB_CHARSET_INFO * c = proxysql_find_charset_nr(mysql_thread___default_charset);
+	if (myds && myds->sess) {
+		csname = myds->sess->mysql_variables->client_get_value(SQL_CHARACTER_SET);
+	}
+
+	const MARIADB_CHARSET_INFO * c = NULL;
+	if (csname)
+		c = proxysql_find_charset_nr(atoi(csname));
+	else
+		c = proxysql_find_charset_name(mysql_thread___default_variables[SQL_CHARACTER_SET]);
+
 	if (!c) {
-		proxy_error("Not existing charset number %u\n", mysql_thread___default_charset);
+		proxy_error("Not existing charset number %s\n", mysql_thread___default_variables[SQL_CHARACTER_SET]);
 		assert(0);
 	}
-	set_charset(c->nr, NAMES);
+	proxy_warning("TRACE : INITIAL ACTION client %s, server %s\n", myds->sess->mysql_variables->client_get_value(SQL_CHARACTER_ACTION), myds->sess->mysql_variables->server_get_value(SQL_CHARACTER_ACTION));
+	set_charset(c->nr, (charset_action)atoi(myds->sess->mysql_variables->client_get_value(SQL_CHARACTER_ACTION)));
+	//set_charset(c->nr, CONNECT_START);
 	mysql_options(mysql, MYSQL_SET_CHARSET_NAME, c->csname);
```

Comment:
Note to myself: the client (proxysql) send to the server (the backend) the default collation for a specific charset).

https://dev.mysql.com/doc/refman/8.0/en/charset-connection.html#charset-connection-client-configuration

If the client is using a collation that is not the default, we may need to change it later on.
We can probably force the desired collation directly into the struct `mysql` , instead of using `mysql_options`